

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>teilab.normalizations &mdash; UiTei-Lab&#39;s Course Documentation  documentation</title>
  

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/plot_directive.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/popup.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="../../_static/favicon.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script src="../../_static/custom.js"></script>
        <script src="../../_static/particles.min.js"></script>
        <script src="../../_static/popup.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> UiTei-Lab's Course Documentation
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
  <ul class="utility-btns">
    <li><a class="btn btn-icon btn-start" href="#" onclick="changeParticles(this);">Start Particles</a></li>
    <li><a class="btn btn-icon btn-expand" href="#" onclick="changeStyle(this);">Expand</a></li>
  </ul>
  <div class="sidebarlogo">
    <h4>Lab's information</h4>
    <ul class="social">
      
      <li><a class="sc-book" href="http://ui-tei.rnai.jp/">
        <i class="fa fa-book" aria-hidden="true"></i>
      </a></li>
      
      <li><a class="sc-wikipedia-w" href="http://ui-tei.rnai.jp/microarray/doku.php?id=2021">
        <i class="fa fa-wikipedia-w" aria-hidden="true"></i>
      </a></li>
      
    </ul>
    <h4>Author's Social link</h4>
    <ul class="social">
      
      <li><a class="sc-twitter" href="https://twitter.com/cabernet_rock">
        <i class="fa fa-twitter" aria-hidden="true"></i>
      </a></li>
      
      <li><a class="sc-github" href="https://github.com/iwasakishuto">
        <i class="fa fa-github" aria-hidden="true"></i>
      </a></li>
      
    </ul>
  </div>
  
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../teilab.html">teilab package</a></li>
</ul>

            
          
        </div>
        

      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">UiTei-Lab's Course Documentation</a>
        
      </nav>


      <div class="wy-nav-content">
<canvas class="particle-background"></canvas>

        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="../index.html">Module code</a> &raquo;</li>
        
      <li>teilab.normalizations</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <h1>Source code for teilab.normalizations</h1><div class="highlight"><pre>
<span></span><span class="c1">#coding: utf-8</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">Differential gene expression can be an outcome of true biological variability or experimental artifacts. Normalization techniques have been used to minimize the effect of experimental artifacts on differential gene expression analysis.</span>

<span class="sd">###############################</span>
<span class="sd">Robust Multichip Analysis (RMA)</span>
<span class="sd">###############################</span>

<span class="sd">In microarray analysis, many algorithms have been proposed, but the most widely used one (**de facto standard**) is :fa:`file-pdf-o` `Robust Multichip Analysis (RMA) &lt;https://academic.oup.com/biostatistics/article/4/2/249/245074&gt;`_ , where the signal value of each spot ( ``RawData`` ) is processed and normalized according to the following flow. ( :ref:`1) Background Subtraction &lt;target to background subtraction section&gt;`, :ref:`2) Normalization Between Samples &lt;target to normalization between samples section&gt;` and :ref:`3) Summarization &lt;target to summarization section&gt;` )</span>

<span class="sd">.. graphviz:: _graphviz/RobustMultichipAnalysis.dot</span>
<span class="sd">      :class: popup-img</span>

<span class="sd">.. _target to background subtraction section:</span>

<span class="sd">*************************</span>
<span class="sd">1. Background Subtraction</span>
<span class="sd">*************************</span>

<span class="sd">In Background Subtraction, we assume that the observed signal intensity is a combnation of the actual signal intensity and the background signal intensity (derived from Non-specific Hybridization), and aim to eliminate the influence of the latter one.</span>

<span class="sd">In the general method, we make assumptions such as</span>

<span class="sd">- Actual signal intensity distribution is an exponential distribution.</span>
<span class="sd">- Background signal intensity distribution is a normal distribution.</span>

<span class="sd">Then, by optimizing the parameters to best represent the phenomenon, the background intensity distribution is calculated and subtracted.</span>

<span class="sd">.. _target to normalization between samples section:</span>

<span class="sd">********************************</span>
<span class="sd">2. Normalization Between Samples</span>
<span class="sd">********************************</span>

<span class="sd">Here, we perform normalization **&quot;between&quot;** samples. What can be said from the results of a **&quot;single sample&quot;** microarray experiment are very limited, and we should compare with other (treatment sample, control group, etc.) experimental results. Howevet, bias due to experimental operation and equipment characteristics is inevitable, and if you just compare them as they are, you will misinterpret them.</span>

<span class="sd">For example, if you want to characterize the changes in global gene expression in the livers of H1/siRNAinsulin-CMV/hIDE transgenic (Tg) mice in response to the reduced bioavailability of insulin [#ref1]_, and the expression level of each RNA in Tg mice was generally lower than that of non-Tg mice, **you may mistakenly conclude that almost all of the RNAs were down-regulated respectively by reduced bioavailability of insulin.**</span>

<span class="sd">.. [#ref1] :fa:`file-pdf-o` `Microarray analysis of insulin-regulated gene expression in the liver: the use of transgenic mice co-expressing insulin-siRNA and human IDE as an animal model &lt;https://pubmed.ncbi.nlm.nih.gov/17982690/&gt;`_</span>

<span class="sd">Therefore, it is necessary to reduce the influence of the bias. There are numerous proposals for normalizing unbalanced data between samples (:fa:`file-pdf-o` `This review paper &lt;https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6988798/&gt;`_ summarizes 23 normalization methods developed for unbalanced transcriptome data), but each method makes some assumptions in data, so it is important to choose the correct normalization method for each experimental results.</span>

<span class="sd">We will introduce two majour methods.</span>

<span class="sd">.. note::</span>

<span class="sd">    1. :ref:`Percentile &lt;target to percentile section&gt;`</span>
<span class="sd">    2. :ref:`Quantile &lt;target to quantile section&gt;`</span>

<span class="sd">.. _target to percentile section:</span>

<span class="sd">1. Percentile</span>
<span class="sd">=============</span>

<span class="sd">This method is a constant adjustoment and the most straightforward.</span>

<span class="sd">1. Calculate the x%tile for each distribution.</span>
<span class="sd">2. Average them.</span>
<span class="sd">3. Divide each distribution by its x%tile and multiply by averaged value.</span>

<span class="sd">+--------------------------------------------------+</span>
<span class="sd">|                    Example                       |</span>
<span class="sd">+==================================================+</span>
<span class="sd">| .. image:: _images/normalizations.percentile.jpg |</span>
<span class="sd">|    :class: popup-img                             |</span>
<span class="sd">+--------------------------------------------------+</span>

<span class="sd">Defined as :func:`percentile &lt;teilab.normalizations.percentile&gt;` in this package.</span>

<span class="sd">.. _target to quantile section:</span>

<span class="sd">2. Quantile</span>
<span class="sd">===========</span>

<span class="sd">Quantile Normalization is a technique for making all distributions identical in statistical properties. It was introduced as **&quot;quantile standardization&quot;** (in :fa:`file-pdf-o` `Analysis of Data from Viral DNA Microchips &lt;https://doi.org/10.1198%2F016214501753381814&gt;`_ ) and then renamed as **&quot;quantile normalization&quot;** (in :fa:`file-pdf-o` `A comparison of normalization methods for high density oligonucleotide array data based on variance and bias &lt;https://doi.org/10.1093%2Fbioinformatics%2F19.2.185&gt;`_ )</span>

<span class="sd">To quantile normalize the all distributions, </span>

<span class="sd">1. Sort each distribution.</span>
<span class="sd">2. Average the data in the same rank.</span>
<span class="sd">3. Replace the value of each rank with the averaged value.</span>

<span class="sd">.. warning::</span>

<span class="sd">    This method can be used when the assumption that &quot;the intensity distribution of gene expression in each sample is almost the same&quot; holds.</span>

<span class="sd">+------------------------------------------------+</span>
<span class="sd">|                    Example                     |</span>
<span class="sd">+================================================+</span>
<span class="sd">| .. image:: _images/normalizations.quantile.jpg |</span>
<span class="sd">|    :class: popup-img                           |</span>
<span class="sd">+------------------------------------------------+</span>

<span class="sd">Defined as :func:`quantile &lt;teilab.normalizations.quantile&gt;` in this package.</span>

<span class="sd">.. _target to summarization section:</span>

<span class="sd">****************</span>
<span class="sd">3. Summarization</span>
<span class="sd">****************</span>

<span class="sd">https://github.com/scipy/scipy/blob/v1.6.3/scipy/signal/signaltools.py#L3384-L3467</span>

<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">scipy.stats.mstats</span> <span class="kn">import</span> <span class="n">gmean</span>
<span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>

<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Any</span>
<span class="kn">from</span> <span class="nn">nptyping</span> <span class="kn">import</span> <span class="n">NDArray</span>
<span class="kn">from</span> <span class="nn">pandas.core.generic</span> <span class="kn">import</span> <span class="n">NDFrame</span>
<span class="kn">from</span> <span class="nn">numbers</span> <span class="kn">import</span> <span class="n">Number</span>


<div class="viewcode-block" id="percentile"><a class="viewcode-back" href="../../teilab.normalizations.html#teilab.normalizations.percentile">[docs]</a><span class="k">def</span> <span class="nf">percentile</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">],</span> <span class="n">percent</span><span class="p">:</span><span class="n">Number</span><span class="o">=</span><span class="mi">75</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Perform Percentile Normalization.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (NDArray[(Any,Any),Number]) : Input data. Shape = ( ``n_samples``, ``n_features`` )</span>
<span class="sd">        percent (Number, optional)       : Which percentile value to normalize. Defaults to ``75``.</span>

<span class="sd">    Returns:</span>
<span class="sd">        NDArray[(Any,Any),Number] : percentiled data. Shape = ( ``n_samples``, ``n_features`` )</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: When ``percent`` tiles contain negative values.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; import matplotlib.pyplot as plt</span>
<span class="sd">        &gt;&gt;&gt; from teilab.normalizations import percentile</span>
<span class="sd">        &gt;&gt;&gt; from teilab.plot.matplotlib import densityplot</span>
<span class="sd">        &gt;&gt;&gt; n_samples, n_features = (4, 1000)</span>
<span class="sd">        &gt;&gt;&gt; data = np.random.RandomState(0).normal(loc=np.expand_dims(np.arange(n_samples), axis=1),  size=(n_samples,n_features))</span>
<span class="sd">        &gt;&gt;&gt; data_percentiled = percentile(data=data, percent=75)</span>
<span class="sd">        &gt;&gt;&gt; fig, axes = plt.subplots(ncols=2, nrows=1, figsize=(12,4))</span>
<span class="sd">        &gt;&gt;&gt; ax = densityplot(data=data, title=&quot;Before Percentile&quot;, ax=axes[0])</span>
<span class="sd">        &gt;&gt;&gt; ax = densityplot(data=data_percentiled, title=&quot;After Percentile&quot;, ax=axes[1])</span>
<span class="sd">    </span>
<span class="sd">    +--------------------------------------------------+</span>
<span class="sd">    |                      Results                     |</span>
<span class="sd">    +==================================================+</span>
<span class="sd">    | .. image:: _images/normalizations.percentile.jpg |</span>
<span class="sd">    |    :class: popup-img                             |</span>
<span class="sd">    +--------------------------------------------------+</span>
<span class="sd">    &quot;&quot;&quot;</span>    
    <span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">percentile</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">q</span><span class="o">=</span><span class="n">percent</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">a</span><span class="o">&lt;</span><span class="mi">0</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">val</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">a</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">val</span><span class="o">&lt;</span><span class="mi">0</span><span class="p">:</span> <span class="k">break</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Geometric mean cannot be calculated because the </span><span class="si">{</span><span class="n">percent</span><span class="si">}</span><span class="s2">%tiles contain negative values (ex. </span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">-th data&#39;s </span><span class="si">{</span><span class="n">percent</span><span class="si">}</span><span class="s2">%tile = </span><span class="si">{</span><span class="n">val</span><span class="si">}</span><span class="s2"> &lt; 0) &quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">data</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">expand_dims</span><span class="p">(</span><span class="n">gmean</span><span class="p">(</span><span class="n">a</span><span class="p">)</span><span class="o">/</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span></div>

<div class="viewcode-block" id="quantile"><a class="viewcode-back" href="../../teilab.normalizations.html#teilab.normalizations.quantile">[docs]</a><span class="k">def</span> <span class="nf">quantile</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Perform Quantile Normalization.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (NDArray[(Any,Any),Number]) : Input data. Shape = ( ``n_samples``, ``n_features`` )</span>

<span class="sd">    Returns:</span>
<span class="sd">        NDArray[(Any,Any),Number] : percentiled data. Shape = ( ``n_samples``, ``n_features`` )</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: When ``data`` contains negative values.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; import matplotlib.pyplot as plt</span>
<span class="sd">        &gt;&gt;&gt; from teilab.normalizations import quantile</span>
<span class="sd">        &gt;&gt;&gt; from teilab.plot.matplotlib import densityplot</span>
<span class="sd">        &gt;&gt;&gt; n_samples, n_features = (4, 1000)</span>
<span class="sd">        &gt;&gt;&gt; data = np.random.RandomState(0).normal(loc=np.expand_dims(np.arange(n_samples), axis=1), size=(n_samples,n_features), ) + 3.5</span>
<span class="sd">        &gt;&gt;&gt; data_quantiled = quantile(data=data)</span>
<span class="sd">        &gt;&gt;&gt; fig, axes = plt.subplots(ncols=2, nrows=1, figsize=(12,4))</span>
<span class="sd">        &gt;&gt;&gt; ax = densityplot(data=data, title=&quot;Before Quantile&quot;, ax=axes[0])</span>
<span class="sd">        &gt;&gt;&gt; ax = densityplot(data=data_quantiled, title=&quot;After Quantile&quot;, ax=axes[1])</span>

<span class="sd">    +------------------------------------------------+</span>
<span class="sd">    |                      Results                   |</span>
<span class="sd">    +================================================+</span>
<span class="sd">    | .. image:: _images/normalizations.quantile.jpg |</span>
<span class="sd">    |    :class: popup-img                           |</span>
<span class="sd">    +------------------------------------------------+</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">any</span><span class="p">(</span><span class="n">data</span><span class="o">&lt;</span><span class="mi">0</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]):</span>
                <span class="k">if</span> <span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span> <span class="o">&lt;</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">break</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Geometric mean cannot be calculated because ``data`` contain negative values. (ex. data[</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">][</span><span class="si">{</span><span class="n">j</span><span class="si">}</span><span class="s2">] = </span><span class="si">{</span><span class="n">data</span><span class="p">[</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">]</span><span class="si">}</span><span class="s2"> &lt; 0)&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gmean</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)[</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">argsort</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)]</span></div>

<div class="viewcode-block" id="median_polish"><a class="viewcode-back" href="../../teilab.normalizations.html#teilab.normalizations.median_polish">[docs]</a><span class="k">def</span> <span class="nf">median_polish</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">],</span> <span class="n">labels</span><span class="p">:</span><span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">),</span><span class="n">Any</span><span class="p">],</span> <span class="n">rtol</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1e-05</span><span class="p">,</span> <span class="n">atol</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1e-08</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">NDArray</span><span class="p">[(</span><span class="n">Any</span><span class="p">,</span><span class="n">Any</span><span class="p">),</span><span class="n">Number</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Median Polish</span>

<span class="sd">    Args:</span>
<span class="sd">        data (NDArray[(Any,Any),Number]) : Input data. Shape = ( ``n_samples``, ``n_features`` )</span>
<span class="sd">        labels (NDArray[(Any),Any])      : Label (ex. ``GeneName``, or ``SystematicName`` )</span>
<span class="sd">        rtol (float)                     : The relative tolerance parameter. Defaults to ``1e-05``</span>
<span class="sd">        atol (float)                     : The absolute tolerance parameter. Defaults to ``1e-08``</span>

<span class="sd">    Raises:</span>
<span class="sd">        TypeError: When ``data.shape[1]`` is not the same as ``len(labels)``</span>

<span class="sd">    Returns:</span>
<span class="sd">        NDArray[(Any,Any),Number]: Median Polished Data.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; from teilab.normalizations import median_polish</span>
<span class="sd">        &gt;&gt;&gt; data = np.asarray([</span>
<span class="sd">        ...     [16.1, 14.6, 19.6, 13.6, 13.6, 13.6],</span>
<span class="sd">        ...     [ 9.0, 18.4,  6.7, 11.1,  6.7,  9.0],</span>
<span class="sd">        ...     [22.4, 13.6, 22.4,  6.7,  9.0,  3.0],</span>
<span class="sd">        &gt;&gt;&gt; ], dtype=float).T</span>
<span class="sd">        &gt;&gt;&gt; n_samples, n_features = data.shape</span>
<span class="sd">        &gt;&gt;&gt; # This means that &quot;All spots (features) are for the same gene.&quot;</span>
<span class="sd">        &gt;&gt;&gt; labels = np.zeros(shape=n_features, dtype=np.uint8)</span>
<span class="sd">        &gt;&gt;&gt; median_polish(data, labels=labels)</span>
<span class="sd">        array([[16.1 , 11.25, 13.3 ],</span>
<span class="sd">               [16.4 , 11.55, 13.6 ],</span>
<span class="sd">               [19.6 , 14.75, 16.8 ],</span>
<span class="sd">               [13.6 ,  8.75, 10.8 ],</span>
<span class="sd">               [11.8 ,  6.95,  9.  ],</span>
<span class="sd">               [13.6 ,  8.75, 10.8 ]])</span>

<span class="sd">    TODO:</span>
<span class="sd">        Speed-UP using ``joblib``, ``multiprocessing``, ``Cython``, etc.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">n_samples</span><span class="p">,</span> <span class="n">n_features</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span>
    <span class="k">if</span> <span class="n">n_features</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;data.shape[1] must be the same as len(labels), but got </span><span class="si">{</span><span class="n">n_features</span><span class="si">}</span><span class="s2">!=</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">labels</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">label</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">),</span> <span class="n">desc</span><span class="o">=</span><span class="s2">&quot;median polish&quot;</span><span class="p">):</span>
        <span class="n">ith_data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[:,</span><span class="n">labels</span><span class="o">==</span><span class="n">label</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">ith_data</span><span class="p">)</span><span class="o">&gt;</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">ith_data_original</span> <span class="o">=</span> <span class="n">ith_data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>    
                <span class="n">feature_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">ith_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
                <span class="n">ith_data</span> <span class="o">=</span> <span class="n">ith_data</span> <span class="o">-</span> <span class="n">feature_median</span><span class="p">[</span><span class="kc">None</span><span class="p">,:]</span>
                <span class="n">sample_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">ith_data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
                <span class="n">ith_data</span> <span class="o">=</span> <span class="n">ith_data</span> <span class="o">-</span> <span class="n">sample_median</span><span class="p">[:,</span><span class="kc">None</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">feature_median</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">,</span> <span class="n">rtol</span><span class="o">=</span><span class="n">rtol</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">sample_median</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">,</span> <span class="n">rtol</span><span class="o">=</span><span class="n">rtol</span><span class="p">):</span>
                    <span class="k">break</span>
            <span class="n">data</span><span class="p">[:,</span><span class="n">labels</span><span class="o">==</span><span class="n">label</span><span class="p">]</span> <span class="o">=</span> <span class="n">ith_data_original</span> <span class="o">-</span> <span class="n">ith_data</span>
    <span class="k">return</span> <span class="n">data</span></div>

<div class="viewcode-block" id="median_polish_group_wise"><a class="viewcode-back" href="../../teilab.normalizations.html#teilab.normalizations.median_polish_group_wise">[docs]</a><span class="k">def</span> <span class="nf">median_polish_group_wise</span><span class="p">(</span><span class="n">data</span><span class="p">:</span><span class="n">NDFrame</span><span class="p">,</span> <span class="n">rtol</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1e-05</span><span class="p">,</span> <span class="n">atol</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1e-08</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">NDFrame</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Apply Median polish group-wise.</span>

<span class="sd">    Args:</span>
<span class="sd">        data (NDFrame)         : Input data. Shape = ( ``n_samples``, ``n_features`` )</span>
<span class="sd">        rtol (float, optional) : [description]. Defaults to ``1e-05``.</span>
<span class="sd">        atol (float, optional) : [description]. Defaults to ``1e-08``.</span>

<span class="sd">    Returns:</span>
<span class="sd">        NDFrame: Median Polished Data.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; from teilab.normalizations import median_polish_group_wise</span>
<span class="sd">        &gt;&gt;&gt; data = pd.DataFrame(data=[</span>
<span class="sd">        ...     [&quot;vimentin&quot;, 16.1, 14.6, 19.6, 13.6, 13.6, 13.6],</span>
<span class="sd">        ...     [&quot;vimentin&quot;,  9.0, 18.4,  6.7, 11.1,  6.7,  9.0],</span>
<span class="sd">        ...     [&quot;vimentin&quot;, 22.4, 13.6, 22.4,  6.7,  9.0,  3.0],</span>
<span class="sd">        &gt;&gt;&gt; ], columns=[&quot;GeneName&quot;]+[f&quot;Samle.{i}&quot; for i in range(6)])</span>
<span class="sd">        &gt;&gt;&gt; data.groupby(&quot;GeneName&quot;).apply(func=median_polish_group_wise).values</span>
<span class="sd">        array([[16.1 , 16.4 , 19.6 , 13.6 , 11.8 , 13.6 ],</span>
<span class="sd">               [11.25, 11.55, 14.75,  8.75,  6.95,  8.75],</span>
<span class="sd">               [13.3 , 13.6 , 16.8 , 10.8 ,  9.  , 10.8 ]])</span>
<span class="sd">        &gt;&gt;&gt; # If you want to see the progress, use tqdm.</span>
<span class="sd">        &gt;&gt;&gt; from tqdm import tqdm</span>
<span class="sd">        &gt;&gt;&gt; tqdm.pandas()</span>
<span class="sd">        &gt;&gt;&gt; data.groupby(&quot;GeneName&quot;).progress_apply(func=median_polish_group_wise).values</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data_original</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="k">while</span> <span class="kc">True</span><span class="p">:</span>
        <span class="n">sample_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span> <span class="o">-</span> <span class="n">sample_median</span><span class="p">[:,</span><span class="kc">None</span><span class="p">]</span>
        <span class="n">feature_median</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span> <span class="o">-</span> <span class="n">feature_median</span><span class="p">[</span><span class="kc">None</span><span class="p">,:]</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">feature_median</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">,</span> <span class="n">rtol</span><span class="o">=</span><span class="n">rtol</span><span class="p">)</span> <span class="ow">and</span> <span class="n">np</span><span class="o">.</span><span class="n">allclose</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="n">sample_median</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">atol</span><span class="o">=</span><span class="n">atol</span><span class="p">,</span> <span class="n">rtol</span><span class="o">=</span><span class="n">rtol</span><span class="p">):</span>
            <span class="k">break</span>
    <span class="k">return</span> <span class="n">data_original</span> <span class="o">-</span> <span class="n">data</span></div>
</pre></div>

           </div>
           
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, iwasakishuto.

    </p>
  </div> 

</footer>
        </div>

      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 
<style id="style-expand" disabled="true">
  .wy-nav-content {
    max-width: 100%!important;
  }
</style>
<script>
  document.getElementById("style-expand").disabled = true;
</script>


</body>
</html>