

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta property="og:title" content="teilab.normalizations module" />
  <meta property="og:type" content="website" />
  <meta property="og:url" content="teilab.normalizations.html" />
  <meta property="og:description" content="This submodule contains various functions and classes that are useful for normalization. Instructions: Differential gene expression can be an outcome of true biological variability or experimental ..." />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>teilab.normalizations module &mdash; UiTei-Lab&#39;s Course Documentation  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/graphviz.css" type="text/css" />
  <link rel="stylesheet" href="_static/plot_directive.css" type="text/css" />
  <link rel="stylesheet" href="_static/togglebutton.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  <link rel="stylesheet" href="_static/popup.css" type="text/css" />

  
  
    <link rel="shortcut icon" href="_static/favicon.png"/>
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/togglebutton.js"></script>
        <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
        <script src="_static/custom.js"></script>
        <script src="_static/particles.min.js"></script>
        <script src="_static/popup.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="teilab.question module" href="teilab.question.html" />
    <link rel="prev" title="teilab.datasets module" href="teilab.datasets.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> UiTei-Lab's Course Documentation
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
  <ul class="utility-btns">
    <li><a class="btn btn-icon btn-start" href="#" onclick="changeParticles(this);">Start Particles</a></li>
    <li><a class="btn btn-icon btn-expand" href="#" onclick="changeStyle(this);">Expand</a></li>
  </ul>
  <div class="sidebarlogo">
    <h4>Lab's information</h4>
    <ul class="social">
      
      <li><a class="sc-book" href="http://ui-tei.rnai.jp/">
        <i class="fa fa-book" aria-hidden="true"></i>
      </a></li>
      
      <li><a class="sc-wikipedia-w" href="http://ui-tei.rnai.jp/microarray/doku.php?id=2021">
        <i class="fa fa-wikipedia-w" aria-hidden="true"></i>
      </a></li>
      
    </ul>
    <h4>Author's Social link</h4>
    <ul class="social">
      
      <li><a class="sc-twitter" href="https://twitter.com/cabernet_rock">
        <i class="fa fa-twitter" aria-hidden="true"></i>
      </a></li>
      
      <li><a class="sc-github" href="https://github.com/iwasakishuto">
        <i class="fa fa-github" aria-hidden="true"></i>
      </a></li>
      
    </ul>
  </div>
  
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="reference internal" href="teilab.html">teilab package</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="teilab.html#subpackages">Subpackages</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="teilab.html#submodules">Submodules</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="teilab.datasets.html">teilab.datasets module</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">teilab.normalizations module</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#instructions">Instructions</a></li>
<li class="toctree-l4"><a class="reference internal" href="#python-objects">Python Objects</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="teilab.question.html">teilab.question module</a></li>
<li class="toctree-l3"><a class="reference internal" href="teilab.seedmatch.html">teilab.seedmatch module</a></li>
<li class="toctree-l3"><a class="reference internal" href="teilab.statistics.html">teilab.statistics module</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="teilab.html#module-teilab">Module contents</a></li>
</ul>
</li>
</ul>

            
          
        </div>
        

      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">UiTei-Lab's Course Documentation</a>
        
      </nav>


      <div class="wy-nav-content">
<canvas class="particle-background"></canvas>

        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
          <li><a href="teilab.html">teilab package</a> &raquo;</li>
        
      <li>teilab.normalizations module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/teilab.normalizations.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-teilab.normalizations">
<span id="teilab-normalizations-module"></span><h1>teilab.normalizations module<a class="headerlink" href="#module-teilab.normalizations" title="Permalink to this headline">¶</a></h1>
<p>This submodule contains various functions and classes that are useful for normalization.</p>
<div class="section" id="instructions">
<h2>Instructions<a class="headerlink" href="#instructions" title="Permalink to this headline">¶</a></h2>
<p>Differential gene expression can be an outcome of true biological variability or experimental artifacts. Normalization techniques have been used to minimize the effect of experimental artifacts on differential gene expression analysis.</p>
<div class="section" id="robust-multichip-analysis-rma">
<h3>Robust Multichip Analysis (RMA)<a class="headerlink" href="#robust-multichip-analysis-rma" title="Permalink to this headline">¶</a></h3>
<p>In microarray analysis, many algorithms have been proposed, but the most widely used one (<strong>de facto standard</strong>) is <em class="fa fa-file-pdf-o"></em> <a class="reference external" href="https://academic.oup.com/biostatistics/article/4/2/249/245074">Robust Multichip Analysis (RMA)</a> , where the signal value of each spot ( <code class="docutils literal notranslate"><span class="pre">RawData</span></code> ) is processed and normalized according to the following flow. ( <a class="reference internal" href="#target-to-background-subtraction-section"><span class="std std-ref">1) Background Subtraction</span></a>, <a class="reference internal" href="#target-to-normalization-between-samples-section"><span class="std std-ref">2) Normalization Between Samples</span></a> and <a class="reference internal" href="#target-to-summarization-section"><span class="std std-ref">3) Summarization</span></a> )</p>
<div class="graphviz"><img src="_images/graphviz-7d9910d4a36548f0df6bfc64bf2016bd419a5dbf.png" alt="digraph RAMPreprocessingGraph {
    graph [
        charset   = &quot;UTF-8&quot;;
        label     = &quot;Preprocessing (RMA)&quot;,
        labelloc  = &quot;t&quot;,
        labeljust = &quot;c&quot;,
        bgcolor   = &quot;#1f441e&quot;,
        fontcolor = &quot;white&quot;,
        fontsize  = 18,
        style     = &quot;filled&quot;,
        rankdir   = TB,
        margin    = 0.2,
        ranksep   = 1.0,
        nodesep   = 0.9,
        layout    = dot,
        compound = true,
    ];

    node [
        style     = &quot;solid,filled&quot;,
        fontsize  = 16,
        fontcolor = 6,
        fontname  = &quot;monospace&quot;,
        color     = &quot;#cee6b4&quot;,
        fillcolor = &quot;#9ecca4&quot;,
        fixedsize = false,
        margin    = &quot;0.2,0.1&quot;,
    ];

    RawData_row1_col1 [shape=doublecircle margin=&quot;0&quot; fillcolor=&quot;#29bb89&quot; fontcolor=&quot;#be0000&quot; color=&quot;#29bb89&quot; label=&quot;RawData\n(row1,col1)&quot;];
    RawData_row1_col2 [shape=doublecircle margin=&quot;0&quot; fillcolor=&quot;#29bb89&quot; fontcolor=&quot;#be0000&quot; color=&quot;#29bb89&quot; label=&quot;RawData\n(row1,col2)&quot;];
    RawData_rowN_colM [shape=doublecircle margin=&quot;0&quot; fillcolor=&quot;#29bb89&quot; fontcolor=&quot;#be0000&quot; color=&quot;#29bb89&quot; label=&quot;RawData\n(rowN,colM)&quot;];
    Ave_row1_col1 [shape=circle margin=&quot;0&quot; fontcolor=&quot;#233e8b&quot; fillcolor=&quot;#b6c9f0&quot; color=&quot;#233e8b&quot; label=&quot;ave&quot;];
    Ave_row1_col2 [shape=circle margin=&quot;0&quot; fontcolor=&quot;#233e8b&quot; fillcolor=&quot;#b6c9f0&quot; color=&quot;#233e8b&quot; label=&quot;ave&quot;];
    Ave_rowN_colM [shape=circle margin=&quot;0&quot; fontcolor=&quot;#233e8b&quot; fillcolor=&quot;#b6c9f0&quot; color=&quot;#233e8b&quot; label=&quot;ave&quot;];
    gMeanSignal_row1_col1 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;110&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gMeanSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;row1,col1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gMeanSignal_row1_col2 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;110&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gMeanSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;row1,col2&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gMeanSignal_rowN_colM [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;110&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gMeanSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;rowN,colM&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gBGUsed_row1_col1 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;80&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gBGUsed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;row1,col1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gBGUsed_row1_col2 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;80&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gBGUsed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;row1,col2&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gBGUsed_rowN_colM [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;white&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;80&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gBGUsed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;location&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;rowN,colM&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gBGSubSignal               [shape=box fontname=&quot;monaco&quot; fontcolor=&quot;#e74c3c&quot; fillcolor=&quot;white&quot; color=&quot;#e1e4e5&quot;];
    rBGSubSignal               [shape=box fontname=&quot;monaco&quot; fontcolor=&quot;#e74c3c&quot; fillcolor=&quot;white&quot; color=&quot;#e1e4e5&quot;];
    gDyeNormSignal             [shape=box fontname=&quot;monaco&quot; fontcolor=&quot;#e74c3c&quot; fillcolor=&quot;white&quot; color=&quot;#e1e4e5&quot;];
    rDyeNormSignal             [shape=box fontname=&quot;monaco&quot; fontcolor=&quot;#e74c3c&quot; fillcolor=&quot;white&quot; color=&quot;#e1e4e5&quot;];

    Minus [shape=circle margin=&quot;0&quot; fontcolor=&quot;#233e8b&quot; fillcolor=&quot;#b6c9f0&quot; color=&quot;#233e8b&quot; label=&quot;-&quot;];

    BackgroundSubtraction          [shape=diamond  fontname=&quot;fantasy&quot; label=&quot;Background Subtraction\n(+ Detrended fluctuation Analysis)&quot;];
    DyeNormalization               [shape=diamond  fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;(Dye Normalization)&quot;];
    SurrogateVariableAnalysis      [shape=diamond  fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;Surrogate Variable Analysis&quot;];
    QuantileNormalization          [shape=diamond  fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;Quantile Normalization&quot;];
    Summarization                  [shape=diamond  fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;Summarization&quot;];
    remaining_processing_row1_col2 [shape=Mdiamond fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;Remaining Processing&quot;]
    remaining_processing_rowN_colM [shape=Mdiamond fontname=&quot;fantasy&quot; margin=&quot;0.05&quot; label=&quot;Remaining Processing&quot;]

    gProcessedSignal_vimentin1 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;160&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignal_insulin1  [
        shape=none 
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;160&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignal_vimentin3 [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;160&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ] 

    gProcessedSignals_sampleA  [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#ffc93c&quot;&gt;
            &lt;tr&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignals&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;vim1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr bgcolor=&quot;red&quot;&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;center&quot; border=&quot;0&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;26&quot; color=&quot;#1f441e&quot;&gt;SAMPLE_A&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;:&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;vim3&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignals_sampleB  [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#ffc93c&quot;&gt;
            &lt;tr&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignals&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;vim1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr bgcolor=&quot;red&quot;&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;center&quot; border=&quot;0&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;26&quot; color=&quot;#1f441e&quot;&gt;SAMPLE_B&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;:&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;vim3&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]

    gProcessedSignal_vimentin1_normed [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;230&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignal_insulin1_normed  [
        shape=none 
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;230&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignal_vimentin3_normed [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;230&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ] 

    gProcessedSignal_vimentin_normed [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#fff5b7&quot;&gt;
            &lt;tr&gt;&lt;td port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;230&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignal_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;gene&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ] 
    gProcessedSignals_sampleB_normed  [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#ffc93c&quot;&gt;
            &lt;tr&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;240&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignals_normed&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;vim1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1_normed&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr bgcolor=&quot;red&quot;&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;center&quot; border=&quot;0&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;26&quot; color=&quot;#1f441e&quot;&gt;SAMPLE_B&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;:&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;vim3&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]
    gProcessedSignals_sampleA_normed  [
        shape = none
        margin = 0
        label = &lt;
        &lt;table width=&quot;100&quot; border=&quot;0&quot; cellspacing=&quot;10&quot; bgcolor=&quot;#ffc93c&quot;&gt;
            &lt;tr&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;left&quot; border=&quot;1&quot; color=&quot;#e1e4e5&quot; bgcolor=&quot;white&quot; width=&quot;240&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;16&quot; face=&quot;monaco&quot; color=&quot;#e74c3c&quot;&gt;gProcessedSignals_normed&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;vim1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin1_normed&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;insulin1_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
            &lt;tr bgcolor=&quot;red&quot;&gt;
            &lt;td rowspan=&quot;2&quot; port=&quot;varname&quot; align=&quot;center&quot; border=&quot;0&quot; width=&quot;170&quot; fixedsize=&quot;true&quot;&gt;&lt;font point-size=&quot;26&quot; color=&quot;#1f441e&quot;&gt;SAMPLE_A&lt;/font&gt;&lt;/td&gt;
            &lt;td port=&quot;ins1&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;:&lt;/font&gt;&lt;/td&gt;
            &lt;/tr&gt;
            &lt;tr&gt;&lt;td port=&quot;vim3&quot;&gt;&lt;font face=&quot;Courier&quot;&gt;vimentin3_normed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
        &lt;/table&gt;
        &gt;
    ]

    edge [
        style         = solid,
        fontsize      = 14,
        fontcolor     = white,
        fontname      = &quot;Migu 1M&quot;,
        color         = white,
        labelfloat    = true,
        labeldistance = 2.5,
        labelangle    = 70
    ];
    RawData_row1_col1 -&gt; Ave_row1_col1 -&gt; gMeanSignal_row1_col1;
    RawData_row1_col2 -&gt; Ave_row1_col2 -&gt; gMeanSignal_row1_col2;
    RawData_rowN_colM -&gt; Ave_rowN_colM -&gt; gMeanSignal_rowN_colM;
    subgraph cluster_1_row1_col1 {
        label     = &quot;Background Subtraction&quot;;
        labelloc  = &quot;t&quot;;
        labeljust = &quot;c&quot;;
        fillcolor = &quot;#89898989&quot;;
        fontcolor = &quot;#ffd56b&quot;;
        margin    = 20;
        gMeanSignal_row1_col1:varname -&gt; BackgroundSubtraction
        BackgroundSubtraction -&gt; gBGUsed_row1_col1;
        gMeanSignal_row1_col1 -&gt; Minus;
        gBGUsed_row1_col1 -&gt; Minus;
        Minus -&gt; gBGSubSignal;
        gBGSubSignal -&gt; DyeNormalization;
        rBGSubSignal -&gt; DyeNormalization;
        DyeNormalization -&gt; gDyeNormSignal;
        DyeNormalization -&gt; rDyeNormSignal;
        gDyeNormSignal -&gt; SurrogateVariableAnalysis;
        SurrogateVariableAnalysis -&gt; gProcessedSignal_vimentin1;
    };
    subgraph cluster_1_row1_col2 {
        label     = &quot;Background Subtraction&quot;;
        labelloc  = &quot;b&quot;;
        labeljust = &quot;l&quot;;
        fillcolor = &quot;#89898989&quot;;
        fontcolor = &quot;#ffd56b&quot;;
        margin    = 20;
        gMeanSignal_row1_col2 -&gt; BackgroundSubtraction -&gt; gBGUsed_row1_col2 -&gt; remaining_processing_row1_col2;
        gMeanSignal_row1_col2 -&gt; remaining_processing_row1_col2;
        remaining_processing_row1_col2 -&gt; gProcessedSignal_insulin1;
    }
    subgraph cluster_1_rowN_colM {
        label     = &quot;Background Subtraction&quot;;
        labelloc  = &quot;b&quot;;
        labeljust = &quot;r&quot;;
        fillcolor = &quot;#89898989&quot;;
        fontcolor = &quot;#ffd56b&quot;;
        margin    = 20;
        gMeanSignal_rowN_colM -&gt; BackgroundSubtraction -&gt; gBGUsed_rowN_colM -&gt; remaining_processing_rowN_colM;
        gMeanSignal_rowN_colM -&gt; remaining_processing_rowN_colM
        remaining_processing_rowN_colM -&gt; gProcessedSignal_vimentin3
    }
    gProcessedSignal_vimentin1 -&gt; gProcessedSignals_sampleA:vim1
    gProcessedSignal_vimentin3 -&gt; gProcessedSignals_sampleA:vim3
    gProcessedSignal_insulin1  -&gt; gProcessedSignals_sampleA:ins1
    subgraph cluster_2 {
        label     = &quot;Normalization Between Samples&quot;;
        labelloc  = &quot;t&quot;;
        labeljust = &quot;l&quot;;
        fillcolor = &quot;#89898989&quot;;
        fontcolor = &quot;#ffd56b&quot;;
        margin    = 20;

        gProcessedSignals_sampleA -&gt; QuantileNormalization;
        gProcessedSignals_sampleB -&gt; QuantileNormalization;
        QuantileNormalization -&gt; gProcessedSignals_sampleB_normed;
        QuantileNormalization -&gt; gProcessedSignals_sampleA_normed;
    };
    gProcessedSignals_sampleA_normed:vim1 -&gt; gProcessedSignal_vimentin1_normed;
    gProcessedSignals_sampleA_normed:ins1 -&gt; gProcessedSignal_insulin1_normed;
    gProcessedSignals_sampleA_normed:vim3 -&gt; gProcessedSignal_vimentin3_normed;
    subgraph cluster_3 {
        label     = &quot;Summarization&quot;;
        labelloc  = &quot;t&quot;;
        labeljust = &quot;l&quot;;
        fillcolor = &quot;#89898989&quot;;
        fontcolor = &quot;#ffd56b&quot;;
        margin    = 20;

        gProcessedSignal_vimentin1_normed -&gt; Summarization;
        gProcessedSignal_vimentin3_normed -&gt; Summarization;
        Summarization -&gt; gProcessedSignal_vimentin_normed;
    };
}" class="graphviz popup-img" /></div>
<div class="section" id="background-subtraction">
<span id="target-to-background-subtraction-section"></span><h4>1. Background Subtraction<a class="headerlink" href="#background-subtraction" title="Permalink to this headline">¶</a></h4>
<p>In Background Subtraction, we assume that the observed signal intensity is a combnation of the actual signal intensity and the background signal intensity (derived from Non-specific Hybridization), and aim to eliminate the influence of the latter one.</p>
<p>In the general method, we make assumptions such as</p>
<ul class="simple">
<li><p>Actual signal intensity distribution is an exponential distribution.</p></li>
<li><p>Background signal intensity distribution is a normal distribution.</p></li>
</ul>
<p>Then, by optimizing the parameters to best represent the phenomenon, the background intensity distribution is calculated and subtracted.</p>
</div>
<div class="section" id="normalization-between-samples">
<span id="target-to-normalization-between-samples-section"></span><h4>2. Normalization Between Samples<a class="headerlink" href="#normalization-between-samples" title="Permalink to this headline">¶</a></h4>
<p>Here, we perform normalization <strong>“between”</strong> samples. What can be said from the results of a <strong>“single sample”</strong> microarray experiment are very limited, and we should compare with other (treatment sample, control group, etc.) experimental results. Howevet, bias due to experimental operation and equipment characteristics is inevitable, and if you just compare them as they are, you will misinterpret them.</p>
<p>For example, if you want to characterize the changes in global gene expression in the livers of H1/siRNAinsulin-CMV/hIDE transgenic (Tg) mice in response to the reduced bioavailability of insulin <a class="footnote-reference brackets" href="#ref1" id="id2">1</a>, and the expression level of each RNA in Tg mice was generally lower than that of non-Tg mice, <strong>you may mistakenly conclude that almost all of the RNAs were down-regulated respectively by reduced bioavailability of insulin.</strong></p>
<dl class="footnote brackets">
<dt class="label" id="ref1"><span class="brackets"><a class="fn-backref" href="#id2">1</a></span></dt>
<dd><p><em class="fa fa-file-pdf-o"></em> <a class="reference external" href="https://pubmed.ncbi.nlm.nih.gov/17982690/">Microarray analysis of insulin-regulated gene expression in the liver: the use of transgenic mice co-expressing insulin-siRNA and human IDE as an animal model</a></p>
</dd>
</dl>
<p>Therefore, it is necessary to reduce the influence of the bias. There are numerous proposals for normalizing unbalanced data between samples (<em class="fa fa-file-pdf-o"></em> <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6988798/">This review paper</a> summarizes 23 normalization methods developed for unbalanced transcriptome data), but each method makes some assumptions in data, so it is important to choose the correct normalization method for each experimental results.</p>
<p>We will introduce two majour methods.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#target-to-percentile-section"><span class="std std-ref">Percentile</span></a></p></li>
<li><p><a class="reference internal" href="#target-to-quantile-section"><span class="std std-ref">Quantile</span></a></p></li>
</ol>
</div>
<div class="section" id="percentile">
<span id="target-to-percentile-section"></span><h5>1. Percentile<a class="headerlink" href="#percentile" title="Permalink to this headline">¶</a></h5>
<p>This method is a constant adjustoment and the most straightforward.</p>
<ol class="arabic simple">
<li><p>Calculate the x%tile for each distribution.</p></li>
<li><p>Average them.</p></li>
<li><p>Divide each distribution by its x%tile and multiply by averaged value.</p></li>
</ol>
<table class="docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><img alt="_images/normalizations.percentile.jpg" class="popup-img" src="_images/normalizations.percentile.jpg" />
</td>
</tr>
</tbody>
</table>
<p>Defined as <a class="reference internal" href="#teilab.normalizations.percentile" title="teilab.normalizations.percentile"><code class="xref py py-func docutils literal notranslate"><span class="pre">percentile</span></code></a> in this package.</p>
</div>
<div class="section" id="quantile">
<span id="target-to-quantile-section"></span><h5>2. Quantile<a class="headerlink" href="#quantile" title="Permalink to this headline">¶</a></h5>
<p>Quantile Normalization is a technique for making all distributions identical in statistical properties. It was introduced as <strong>“quantile standardization”</strong> (in <em class="fa fa-file-pdf-o"></em> <a class="reference external" href="https://doi.org/10.1198%2F016214501753381814">Analysis of Data from Viral DNA Microchips</a> ) and then renamed as <strong>“quantile normalization”</strong> (in <em class="fa fa-file-pdf-o"></em> <a class="reference external" href="https://doi.org/10.1093%2Fbioinformatics%2F19.2.185">A comparison of normalization methods for high density oligonucleotide array data based on variance and bias</a> )</p>
<p>To quantile normalize the all distributions,</p>
<ol class="arabic simple">
<li><p>Sort each distribution.</p></li>
<li><p>Average the data in the same rank.</p></li>
<li><p>Replace the value of each rank with the averaged value.</p></li>
</ol>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method can be used when the assumption that “the intensity distribution of gene expression in each sample is almost the same” holds.</p>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><img alt="_images/normalizations.quantile.jpg" class="popup-img" src="_images/normalizations.quantile.jpg" />
</td>
</tr>
</tbody>
</table>
<p>Defined as <a class="reference internal" href="#teilab.normalizations.quantile" title="teilab.normalizations.quantile"><code class="xref py py-func docutils literal notranslate"><span class="pre">quantile</span></code></a> in this package.</p>
</div>
</div>
<div class="section" id="summarization">
<span id="target-to-summarization-section"></span><h4>3. Summarization<a class="headerlink" href="#summarization" title="Permalink to this headline">¶</a></h4>
<p><a class="reference external" href="https://github.com/scipy/scipy/blob/v1.6.3/scipy/signal/signaltools.py#L3384-L3467">https://github.com/scipy/scipy/blob/v1.6.3/scipy/signal/signaltools.py#L3384-L3467</a></p>
</div>
</div>
</div>
<div class="section" id="python-objects">
<h2>Python Objects<a class="headerlink" href="#python-objects" title="Permalink to this headline">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="teilab.normalizations.percentile">
<span class="sig-prename descclassname"><span class="pre">teilab.normalizations.</span></span><span class="sig-name descname"><span class="pre">percentile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">percent</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">numbers.Number</span></span> <span class="o"><span class="pre">=</span></span> <span class="default_value"><span class="pre">75</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span><a class="reference internal" href="_modules/teilab/normalizations.html#percentile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teilab.normalizations.percentile" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform Percentile Normalization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>NDArray</em><em>[</em><em>(</em><em>Any</em><em>,</em><em>Any</em><em>)</em><em>,</em><em>Number</em><em>]</em>) – Input data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p></li>
<li><p><strong>percent</strong> (<em>Number</em><em>, </em><em>optional</em>) – Which percentile value to normalize. Defaults to <code class="docutils literal notranslate"><span class="pre">75</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>percentiled data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>NDArray[(Any,Any),Number]</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – When <code class="docutils literal notranslate"><span class="pre">percent</span></code> tiles contain negative values.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.normalizations</span> <span class="kn">import</span> <span class="n">percentile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.plot.matplotlib</span> <span class="kn">import</span> <span class="n">densityplot</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n_samples</span><span class="p">,</span> <span class="n">n_features</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">RandomState</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">expand_dims</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_samples</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span>  <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span><span class="n">n_features</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data_percentiled</span> <span class="o">=</span> <span class="n">percentile</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">percent</span><span class="o">=</span><span class="mi">75</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">densityplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Before Percentile&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">densityplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_percentiled</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;After Percentile&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Results</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><img alt="_images/normalizations.percentile.jpg" class="popup-img" src="_images/normalizations.percentile.jpg" />
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teilab.normalizations.quantile">
<span class="sig-prename descclassname"><span class="pre">teilab.normalizations.</span></span><span class="sig-name descname"><span class="pre">quantile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span><a class="reference internal" href="_modules/teilab/normalizations.html#quantile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teilab.normalizations.quantile" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform Quantile Normalization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>data</strong> (<em>NDArray</em><em>[</em><em>(</em><em>Any</em><em>,</em><em>Any</em><em>)</em><em>,</em><em>Number</em><em>]</em>) – Input data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>percentiled data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>NDArray[(Any,Any),Number]</p>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>ValueError</strong> – When <code class="docutils literal notranslate"><span class="pre">data</span></code> contains negative values.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.normalizations</span> <span class="kn">import</span> <span class="n">quantile</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.plot.matplotlib</span> <span class="kn">import</span> <span class="n">densityplot</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n_samples</span><span class="p">,</span> <span class="n">n_features</span> <span class="o">=</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">RandomState</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span><span class="o">.</span><span class="n">normal</span><span class="p">(</span><span class="n">loc</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">expand_dims</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">n_samples</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="n">n_samples</span><span class="p">,</span><span class="n">n_features</span><span class="p">),</span> <span class="p">)</span> <span class="o">+</span> <span class="mf">3.5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data_quantiled</span> <span class="o">=</span> <span class="n">quantile</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">fig</span><span class="p">,</span> <span class="n">axes</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span> <span class="n">nrows</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">densityplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Before Quantile&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ax</span> <span class="o">=</span> <span class="n">densityplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">data_quantiled</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;After Quantile&quot;</span><span class="p">,</span> <span class="n">ax</span><span class="o">=</span><span class="n">axes</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 100%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Results</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><img alt="_images/normalizations.quantile.jpg" class="popup-img" src="_images/normalizations.quantile.jpg" />
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teilab.normalizations.median_polish">
<span class="sig-prename descclassname"><span class="pre">teilab.normalizations.</span></span><span class="sig-name descname"><span class="pre">median_polish</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">Any</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rtol</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">float</span></span> <span class="o"><span class="pre">=</span></span> <span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atol</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">float</span></span> <span class="o"><span class="pre">=</span></span> <span class="default_value"><span class="pre">1e-08</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">nptyping.types._ndarray.NDArray</span><span class="p"><span class="pre">[</span></span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">typing.Any</span><span class="p"><span class="pre">,</span> </span><span class="pre">nptyping.types._object.Object</span><span class="p"><span class="pre">]</span></span><a class="reference internal" href="_modules/teilab/normalizations.html#median_polish"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teilab.normalizations.median_polish" title="Permalink to this definition">¶</a></dt>
<dd><p>Median Polish</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>NDArray</em><em>[</em><em>(</em><em>Any</em><em>,</em><em>Any</em><em>)</em><em>,</em><em>Number</em><em>]</em>) – Input data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p></li>
<li><p><strong>labels</strong> (<em>NDArray</em><em>[</em><em>(</em><em>Any</em><em>)</em><em>,</em><em>Any</em><em>]</em>) – Label (ex. <code class="docutils literal notranslate"><span class="pre">GeneName</span></code>, or <code class="docutils literal notranslate"><span class="pre">SystematicName</span></code> )</p></li>
<li><p><strong>rtol</strong> (<em>float</em>) – The relative tolerance parameter. Defaults to <code class="docutils literal notranslate"><span class="pre">1e-05</span></code></p></li>
<li><p><strong>atol</strong> (<em>float</em>) – The absolute tolerance parameter. Defaults to <code class="docutils literal notranslate"><span class="pre">1e-08</span></code></p></li>
</ul>
</dd>
<dt class="field-even">Raises</dt>
<dd class="field-even"><p><strong>TypeError</strong> – When <code class="docutils literal notranslate"><span class="pre">data.shape[1]</span></code> is not the same as <code class="docutils literal notranslate"><span class="pre">len(labels)</span></code></p>
</dd>
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Median Polished Data.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>NDArray[(Any,Any),Number]</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.normalizations</span> <span class="kn">import</span> <span class="n">median_polish</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">16.1</span><span class="p">,</span> <span class="mf">14.6</span><span class="p">,</span> <span class="mf">19.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span> <span class="mf">9.0</span><span class="p">,</span> <span class="mf">18.4</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span> <span class="mf">11.1</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span>  <span class="mf">9.0</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="mf">22.4</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">22.4</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span>  <span class="mf">9.0</span><span class="p">,</span>  <span class="mf">3.0</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">T</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">n_samples</span><span class="p">,</span> <span class="n">n_features</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">shape</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># This means that &quot;All spots (features) are for the same gene.&quot;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">shape</span><span class="o">=</span><span class="n">n_features</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">uint8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">median_polish</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="n">labels</span><span class="p">)</span>
<span class="go">array([[16.1 , 11.25, 13.3 ],</span>
<span class="go">       [16.4 , 11.55, 13.6 ],</span>
<span class="go">       [19.6 , 14.75, 16.8 ],</span>
<span class="go">       [13.6 ,  8.75, 10.8 ],</span>
<span class="go">       [11.8 ,  6.95,  9.  ],</span>
<span class="go">       [13.6 ,  8.75, 10.8 ]])</span>
</pre></div>
</div>
<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>Speed-UP using <code class="docutils literal notranslate"><span class="pre">joblib</span></code>, <code class="docutils literal notranslate"><span class="pre">multiprocessing</span></code>, <code class="docutils literal notranslate"><span class="pre">Cython</span></code>, etc.</p>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="teilab.normalizations.median_polish_group_wise">
<span class="sig-prename descclassname"><span class="pre">teilab.normalizations.</span></span><span class="sig-name descname"><span class="pre">median_polish_group_wise</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">pandas.core.generic.NDFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rtol</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">float</span></span> <span class="o"><span class="pre">=</span></span> <span class="default_value"><span class="pre">1e-05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atol</span></span><span class="p"><span class="pre">:</span></span> <span class="n"><span class="pre">float</span></span> <span class="o"><span class="pre">=</span></span> <span class="default_value"><span class="pre">1e-08</span></span></em><span class="sig-paren">)</span> &#x2192; <span class="pre">pandas.core.generic.NDFrame</span><a class="reference internal" href="_modules/teilab/normalizations.html#median_polish_group_wise"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#teilab.normalizations.median_polish_group_wise" title="Permalink to this definition">¶</a></dt>
<dd><p>Apply Median polish group-wise.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data</strong> (<em>NDFrame</em>) – Input data. Shape = ( <code class="docutils literal notranslate"><span class="pre">n_samples</span></code>, <code class="docutils literal notranslate"><span class="pre">n_features</span></code> )</p></li>
<li><p><strong>rtol</strong> (<em>float</em><em>, </em><em>optional</em>) – [description]. Defaults to <code class="docutils literal notranslate"><span class="pre">1e-05</span></code>.</p></li>
<li><p><strong>atol</strong> (<em>float</em><em>, </em><em>optional</em>) – [description]. Defaults to <code class="docutils literal notranslate"><span class="pre">1e-08</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Median Polished Data.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>NDFrame</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">teilab.normalizations</span> <span class="kn">import</span> <span class="n">median_polish_group_wise</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="p">[</span>
<span class="gp">... </span>    <span class="p">[</span><span class="s2">&quot;vimentin&quot;</span><span class="p">,</span> <span class="mf">16.1</span><span class="p">,</span> <span class="mf">14.6</span><span class="p">,</span> <span class="mf">19.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="s2">&quot;vimentin&quot;</span><span class="p">,</span>  <span class="mf">9.0</span><span class="p">,</span> <span class="mf">18.4</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span> <span class="mf">11.1</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span>  <span class="mf">9.0</span><span class="p">],</span>
<span class="gp">... </span>    <span class="p">[</span><span class="s2">&quot;vimentin&quot;</span><span class="p">,</span> <span class="mf">22.4</span><span class="p">,</span> <span class="mf">13.6</span><span class="p">,</span> <span class="mf">22.4</span><span class="p">,</span>  <span class="mf">6.7</span><span class="p">,</span>  <span class="mf">9.0</span><span class="p">,</span>  <span class="mf">3.0</span><span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GeneName&quot;</span><span class="p">]</span><span class="o">+</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;Samle.</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">6</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;GeneName&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">median_polish_group_wise</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
<span class="go">array([[16.1 , 16.4 , 19.6 , 13.6 , 11.8 , 13.6 ],</span>
<span class="go">       [11.25, 11.55, 14.75,  8.75,  6.95,  8.75],</span>
<span class="go">       [13.3 , 13.6 , 16.8 , 10.8 ,  9.  , 10.8 ]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># If you want to see the progress, use tqdm.</span>
<span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">tqdm</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tqdm</span><span class="o">.</span><span class="n">pandas</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">data</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;GeneName&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">progress_apply</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">median_polish_group_wise</span><span class="p">)</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</dd></dl>

</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="teilab.question.html" class="btn btn-neutral float-right" title="teilab.question module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
        <a href="teilab.datasets.html" class="btn btn-neutral float-left" title="teilab.datasets module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2021, iwasakishuto.

    </p>
  </div> 

</footer>
        </div>

      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
  
 
<style id="style-expand" disabled="true">
  .wy-nav-content {
    max-width: 100%!important;
  }
</style>
<script>
  document.getElementById("style-expand").disabled = true;
</script>


</body>
</html>